tests/unit/test_dbapi_cursor.py

```diff
@@ -734,3 +747,80 @@ def test__format_operation_unescaped_percent_w_list_param(self):
             "SELECT %s, %s, '100 %';",
             ["foo", "bar"],
         )
+
+    def test__format_operation_no_placeholders(self):
+        from google.cloud.bigquery import dbapi
+        from google.cloud.bigquery.dbapi import cursor
+
+        self.assertRaises(
+            dbapi.ProgrammingError,
+            cursor._format_operation,
+            "SELECT 42",
+            ["foo", "bar"],
+        )
+
+
+@pytest.mark.parametrize(
+    "inp,expect",
+    [
+        ("", ("", None)),
+        ("values(%(foo)s, %(bar)s)", ("values(%(foo)s, %(bar)s)", {})),
+        (
+            "values('%%(oof:INT64)s', %(foo)s, %(bar)s)",
+            ("values('%%(oof:INT64)s', %(foo)s, %(bar)s)", {}),
+        ),
+        (
+            "values(%(foo:INT64)s, %(bar)s)",
+            ("values(%(foo)s, %(bar)s)", dict(foo="INT64")),
+        ),
+        (
+            "values('%%(oof:INT64)s, %(foo:INT64)s, %(foo)s)",
+            ("values('%%(oof:INT64)s, %(foo)s, %(foo)s)", dict(foo="INT64")),
+        ),
+        (
+            "values(%(foo:INT64)s, %(foo:INT64)s)",
```

Comment:
Can we add some tests that check our aliases, too? (INTEGER -> INT64, for example)

Reply:
I think there is a backend alias for `INTEGER`, so it's not strictly necessary, but will be if we want to support aliases like `DECIMAL` in the future.

Reply:
That's included in my latest push.

We do need it  if we want to do client-side checks.